File: dku_bluesky_analysis/Dku_bluesky_analysis/ebs_read_tbe.R
#' Read tbe file
#'
#' By default: All table names will be converted to lower case
#' "Timeseries" table will load into tb (data) and tc (metadata)
#' Other tables will load into eg. tb_sites and tc_sites
#'
#' Returns a list "result" containing data tibbles (tc_...) and associated metadata (tc_...) or an error message
#'
#' Variable name conventions:
#' Prefixes:
#' tf: table read using fread
#'   tf: all
#'   tf_tbl: single table read
#' tb: tibble of data
#' tc: tibble of metadata
#'
#' hdr: header variables
#' att: attribute variables (aka metadata)
#' cmt: comments
#'
#' i...: index of row in the file
#' n...: number of records/rows/attributes
#'
#' @param flin File name
#' @param flsource Source: extdata or directory
#' @param tblselect Null: Read all tables, or name of table to read
#'
#' @return result: List of tibbles with tables and metadata, error: Error Message
#' @export
#'
#' @importFrom data.table fread
#' @importFrom stringr str_sub str_replace str_replace_all str_match
#' @importFrom lubridate force_tz
#' @importFrom dplyr select
#' @importFrom rlang .data

library(tidyverse)
library(lubridate)
library(data.table)

ebs_read_tbe <- function(flin = 'ebs_sites_20220125_tbe.csv',
                         flsource = 'extdata',
                         tblselect = NULL) {
  #flin <- 'site_meta.csv'
  #flsource <- './'
  #tblselect <- NULL # or table name to read, eg. 'seasons'
  debugout <- 0 # debugout > 0: output debug info from reading TBE file

  result = list() # initialize output

  if ( identical(flsource,'extdata')) {
    flin_full <- system.file("extdata", flin, package="Aqsebs")
    if ( identical(flin_full,'') ) {
      return( list(result=NULL, error=sprintf('File not found: %s',flin)) )
    }
  } else {
    flin_full <- paste(flsource,flin,sep='')
  }

  if ( ! file_test('-f',flin_full) ) {
    return( list(result=NULL, error=sprintf('File not found: %s',flin_full)) )
  } else {
    if ( debugout ) { printf('Will read file %s',flin_full) }
  }

  # read and parse header column, read second column to fill in missing EOT:
  tf <- fread(flin_full, header=FALSE, sep=',', select=c(1,2), fill=TRUE)

  #itbl_header <- which(str_detect(a$V1,'^TBL'))
  nrecs <- length(tf$V1)
  tf_codes <- str_sub(tf$V1,1,3)
  tf_description <- str_sub(tf$V1,5,-1)
  itbl_header <- which( tf_codes == 'TBL' )
  ntables = length(itbl_header)

  for ( ntbl in 1:ntables ) {
    iheader = itbl_header[ntbl]
    tbl_str = tolower(tf_description[iheader])
    if ( !is.null(tblselect) && tbl_str != tolower(tblselect) ) {
      if ( debugout ) { printf('Skipping table %s',tbl_str) }
      next
    }

    # get last possible row in this table: either the row preceding the next table, or the end of the file:
    if ( ntbl < ntables) {
      ilast = itbl_header[ntbl+1]-1
    } else {
      ilast = nrecs
    }

    # Find end of data:
    ieot <- which( tf_codes[(iheader+1):ilast]=='EOT' )
    skipdata <- 0
    if ( length(ieot) > 1 ) {
      return( list(result=NULL,
                   error=sprintf('Too many EOT found for table in ebs_read_tbe')) )
    } else if ( length(ieot) == 0 ) {
      inotblank = which( tf$V2[(iheader+1):ilast] != '' & tf$V1[(iheader+1):ilast] == '' )
      if ( length(inotblank) > 0 ) {
        iend <- iheader + max(inotblank)
        if ( debugout ) {
          printf('EOT not found for table %d/%d, using iend = %d',ntbl,ntables,iend)
        }
      } else {
        # look for metadata:
        inotblank2 = which( tf$V2[(iheader+1):ilast] != '' )
        if ( length(inotblank2) > 0 ) {
          iend <- iheader + max(inotblank2)
          if ( debugout ) {
            printf('Found metadata but no data for table %d/%d, using iend = %d',ntbl,ntables,iend)
          }
          skipdata <- 1
        } else {
          if ( debugout ) {
            printf('No data and no metadata found for table %d/%d (consider fixing your file)',ntbl,ntables)
          }
          next
        }
      }
    } else {
      iend <- ieot + iheader
    }

    # Find start of data:
    if ( skipdata ) {
      # set istart to after table so all attributes are read
      istart <- iend + 1
      ndata <- 0
    } else {
        ibgn <- which( tf_codes[(iheader+1):iend]=='BGN' )
    if ( length(ibgn) > 1 ) {
      return( list(result=NULL,
                   error=sprintf('Too many BGN found for table in ebs_read_tbe')) )
    } else if ( length(ibgn) == 0 ) {
      iblank <- which(tf$V1[(iheader+1):iend]=='')
      if ( length(iblank) > 0 ) {
        # Use first row with blank code in the first column:
        istart <- iheader + min(iblank)
      } else {
        # only one row of data signaled by EOT
        istart <- iend
      }
      if ( debugout ) {
        printf('BGN not found for table %d/%d, using istart = %d',ntbl,ntables,istart)
      }
    } else {
      istart <- ibgn + iheader
    }
    ndata <- iend - istart + 1
    }

    # Read column names:
    hdr_all <- fread(flin_full, header=FALSE, sep=',',
                     skip=iheader-1, nrows=1, fill=TRUE)
    tbl_name <- str_replace(hdr_all$V1, '^TBL ', '')
    tbl_name <- str_replace_all(tbl_name, '\\s', '_')
    hdr_select <- which(!is.na(hdr_all))
    hdr <- hdr_all[ , hdr_select, with=FALSE]
    hdr$V1 <- paste('TBL_', tbl_name,sep='')
    hdr_data <- hdr_all[ , hdr_select, with=FALSE]
    #hdr_data<-hdr # when I do this, setting V1:=NULL affects both hdr_data and hdr: bug or feature?
    hdr_data <- select(hdr_data,-1) # alternative to (without raising warnings): hdr_data[,V1:=NULL]

    # Find and read metadata:
    iatt <- which( tf_codes[(iheader+1):(istart-1)]=='ATT' )
    if ( length(iatt) == 0 ) {
      if ( debugout ) {
        printf('No attributes found for Table %d/%d',ntbl,ntables)
      }
      att_trans = NULL
    } else {
      iatt1 = min(iatt) + iheader
      iatt2 = max(iatt) + iheader
      natt = iatt2 - iatt1 + 1
      att <- fread(flin_full, header=FALSE, sep=',',
                   skip=iatt1-1, nrows=natt, select=hdr_select, fill=TRUE)
      att_o <- att
      att <- att[iatt,]
      # In column 1 we have attribute names, remove ATT to keep just the names:
      att$V1 <- str_replace(att$V1,'^ATT ','')
      colnames(att) <- as.character(hdr)
      att_trans <- df_transpose(att)
      colnames(att_trans)[1] <- 'Variable'
    }

    # Read data:
    if ( ndata > 0 ) {
      # Need to read first column to screen for comments:

      tf_tbl <- fread(flin_full, header=FALSE, sep=',',
                      skip=istart-1, nrows=ndata, select=hdr_select, fill=TRUE)
      tf_tbl_codes <- str_sub(tf_tbl$V1,1,3)
      icmt <- ( tf_tbl_codes == "CMT")
      ncmt <- length(which(icmt))
      if ( ncmt > 0 ) {
        if ( debugout ) {
          printf('Removing %d comments from table %s',ncmt,tbl_str)
        }
        tf_tbl <- tf_tbl[!icmt]
      }
      tf_tbl[['V1']] <- NULL
      colnames(tf_tbl) <- as.character(hdr_data)
      # reset timezones for all dates:
      ##for ( var_col in which(sapply(tf_tbl,is.POSIXt)) ) { # does not work because just dates do not get encoded as POSIXt # need to import is.POSIXt from lubridate
      var_list = names(tf_tbl)
      for ( nv in 1:length(var_list) ) {
        vselect = var_list[nv]
        rtn = get_attribute_check(att_trans, vselect, 'Units')
        if ( is.null(rtn$result) ) {
          #printf('Did not find Units for %s, ignore for now (%s)',vselect,rtn$error)
          next
        }
        ## Check if we have a date/time (is.Date did not work, class returns POSIXct and POSIXt:
        if ( pillar::type_sum(tf_tbl[[vselect]]) == 'dttm' ) {
          ## Do we have time variable, units = "xxx Time (Est/UTC) xxx":
          tzone = str_match(rtn$result,"Time\\s*\\(\\s*(.*)\\s*\\)")[2]
          if ( ! is.na(tzone) ) {
            if ( tzone %in% OlsonNames() ) {
              printf('Found valid time and timezone (%s) for column %s',tzone,vselect)
              tf_tbl[[vselect]] <- force_tz(tf_tbl[[vselect]],tzone)
              next
            } else {
              printf('Invalid timezone (%s) for column %s, leave as is',tzone,vselect)
            }
          } else if ( rtn$result %in% OlsonNames() ) {
            tzone = rtn$result
            printf('Inferred time, found timezone (%s) for column %s',tzone,vselect)
              tf_tbl[[vselect]] <- force_tz(tf_tbl[[vselect]],tzone)
              next
          } else {
            printf('WARNING: No timezone found for variable %s, leave as is',vselect)
          }
        } else {
          #printf('Variable %s (%s) not a time variable, leave as is',vselect,rtn$result)
        }
      }
    } else {
      tf_tbl <- NULL
    }

    # Add data to result:
    # Find name for tibble, use "tb" and "tc" for timeseries:
    if ( identical(tbl_str,'timeseries') ) {
      tbname = sprintf('tb')
      tcname = sprintf('tc')
      ntb <- 1
      while ( any(names(result)==tbname) || any(names(result)==tcname) ) {
        ntb <- ntb + 1
        tbname = sprintf('tb%d',ntb)
        tcname = sprintf('tc%d',ntb)
      }
    } else {
      tbname = sprintf('tb_%s',tbl_str)
      tcname = sprintf('tc_%s',tbl_str)
      ntb <- 1
      while ( any(names(result)==tbname) || any(names(result)==tcname) ) {
        ntb <- ntb + 1
        tbname = sprintf('tb_%s%d',tbl_str,ntb)
        tcname = sprintf('tc_%s%d',tbl_str,ntb)
      }
    }

    if ( is.null(tf_tbl) ) {
      result[[tbname]] <- NULL
    } else if ( tolower(tbl_str) == 'global' ) {
      tb_gbl <- df_transpose(tibble(tf_tbl))
      #tb_gbl[[1]] <- NULL # if we don't need variable/value as a pair.
      result[[tbname]] <- tb_gbl
    } else {
      result[[tbname]] <- tibble(tf_tbl)
    }
    if ( is.null(att_trans) ) {
      result[[tcname]] <- NULL
    } else {
      result[[tcname]] <- tibble(att_trans)
    }

    # Summary message:
    printf('Tibbles %s and %s for table %d/%d: %s',tbname, tcname,ntbl,ntables,tf_description[iheader])
    if ( debugout ) {
      if ( ndata > 0 ) {
        printf('  Rows: header at %d, data from %d to %d',iheader,istart,iend)
      } else if ( !is.null(att_trans) ) {
        printf('  Rows: header at %d, metadata only from %d to %d',iheader,iatt1,iatt2)
      } else {
        printf('  Rows: header only at %d, no data, no metadata',iheader)
      }
      printf('  Columns: %s',paste(hdr_select,collapse=', '))
    }
  }
  printf('Finished reading file: %s',flin_full)
  return( list(result=result, error=NULL) )
}

File: python_conv/ebs_read_tbe.py
import pandas as pd
import re
from tzlocal import get_localzone
from pytz import all_timezones

from bdf_utils import get_attribute_check

def ebs_read_tbe(flin='./Dku_bluesky_analysis/saq_bluesky_bgd_20211001_20230430_inv_tbe.csv', flsource='extdata', tblselect=None):
    """
    Read tbe file

    By default: All table names will be converted to lower case
    "Timeseries" table will load into tb (data) and tc (metadata)
    Other tables will load into eg. tb_sites and tc_sites

    Returns a dictionary "result" containing data dataframes (tc_...) and associated metadata (tc_...) or an error message

    Variable name conventions:
    Prefixes:
    tf: table read using pandas read_csv
    tf_tbl: single table read
    tb: dataframe of data
    tc: dataframe of metadata

    hdr: header variables
    att: attribute variables (aka metadata)
    cmt: comments

    i...: index of row in the file
    n...: number of records/rows/attributes

    Parameters:
    - flin: File name (default: '/Users/stuart/Desktop/python_tbe/Dku_bluesky_analysis/saq_bluesky_bgd_20211001_20230430_inv_tbe.csv')
    - flsource: Source: 'extdata' or directory (default: 'extdata')
    - tblselect: Null: Read all tables, or name of table to read

    Returns:
    - result: Dictionary of dataframes with tables and metadata, error: Error Message
    """

    result = {}  # initialize output

    # if flsource == 'extdata':
    #     flin_full = pkg_resources.resource_filename('Aqsebs', f'extdata/{flin}')
    #     if not os.path.exists(flin_full):
    #         return {'result': None, 'error': f'File not found: {flin}'}
    # else:
    #     flin_full = os.path.join(flsource, flin)

    # if not os.path.isfile(flin_full):
    #     return {'result': None, 'error': f'File not found: {flin_full}'}
    # else:
    #     print(f'Will read file {flin_full}')

    # read and parse header column, read second column to fill in missing EOT:
    tf = pd.read_csv(flin, header=None, sep=',', usecols=[0, 1], keep_default_na=False)
    nrecs = len(tf)
    print(tf)
    tf_codes = tf[0].str[:3]
    tf_description = tf[0].str[4:]
    itbl_header = tf_codes[tf_codes == 'TBL'].index
    ntables = len(itbl_header)

    for ntbl in range(1, ntables + 1):
        iheader = itbl_header[ntbl - 1]
        tbl_str = tf_description[iheader].lower()
        if tblselect is not None and tbl_str != tblselect.lower():
            print(f'Skipping table {tbl_str}')
            continue

        # get last possible row in this table: either the row preceding the next table, or the end of the file:
        if ntbl < ntables:
            ilast = itbl_header[ntbl] - 1
        else:
            ilast = nrecs

        # Find end of data:
        ieot = (tf_codes[(iheader + 1):ilast] == 'EOT').idxmax()
        skipdata = False
        if pd.notnull(ieot):
            iend = iheader + ieot
        else:
            inotblank = ((tf[1][(iheader + 1):ilast] != '') & (tf[0][(iheader + 1):ilast] == '')).idxmax()
            if pd.notnull(inotblank):
                iend = iheader + inotblank
                print(f'EOT not found for table {ntbl}/{ntables}, using iend = {iend}')
            else:
                inotblank2 = (tf[1][(iheader + 1):ilast] != '').idxmax()
                if pd.notnull(inotblank2):
                    iend = iheader + inotblank2
                    print(f'Found metadata but no data for table {ntbl}/{ntables}, using iend = {iend}')
                    skipdata = True
                else:
                    print(f'No data and no metadata found for table {ntbl}/{ntables} (consider fixing your file)')
                    continue

        # Find start of data:
        if skipdata:
            # set istart to after table so all attributes are read
            istart = iend + 1
            ndata = 0
        else:
            ibgn = (tf_codes[(iheader + 1):iend] == 'BGN').idxmax()
            if pd.notnull(ibgn):
                istart = iheader + ibgn
            else:
                iblank = (tf[0][(iheader + 1):iend] == '').idxmax()
                if pd.notnull(iblank):
                    # Use first row with blank code in the first column:
                    istart = iheader + iblank
                else:
                    # only one row of data signaled by EOT
                    istart = iend
                    print(f'BGN not found for table {ntbl}/{ntables}, using istart = {istart}')
            ndata = iend - istart + 1

        # Read column names:
        hdr_all = pd.read_csv(flin, header=None, sep=',', skiprows=iheader, nrows=1, keep_default_na=False)
        tbl_name = hdr_all[0].str.replace('^TBL ', '')
        tbl_name = tbl_name.str.replace('\s', '_')
        hdr_select = hdr_all.columns[~hdr_all.isna().any()]
        hdr = hdr_all[hdr_select].reset_index(drop=True)
        hdr[0] = 'TBL_' + tbl_name
        hdr_data = hdr_all[hdr_select].reset_index(drop=True)

        # Find and read metadata:
        itbl_sites = (tf_codes == 'TBL Sites').idxmax()
        ibgn = (tf_codes == 'BGN').idxmax()

        if pd.notnull(itbl_sites) and pd.notnull(ibgn):
            iatt_start = itbl_sites + 1
            iatt_end = ibgn - 1
            att_indices = (tf_codes[iatt_start:iatt_end] == 'ATT')

            if not att_indices.any():
                print(f"No 'ATT' rows found between 'TBL Sites' and 'BGN' for table {ntbl}/{ntables}")
                continue  # Skip to the next table

            iatt = att_indices.idxmax()
        else:
            print(f"'TBL Sites' or 'BGN' row not found for table {ntbl}/{ntables}")
            #continue  # Skip to the next table
            if pd.notnull(iatt):
                iatt1 = iatt + iheader
                iatt2 = iheader + (tf_codes[(iheader + 1):(istart - 1)] == 'ATT').last_valid_index()
                natt = iatt2 - iatt1 + 1
                att = pd.read_csv(flin, header=None, sep=',', skiprows=iatt1, nrows=natt, usecols=hdr_select,
                                keep_default_na=False)
                att.columns = hdr.iloc[0]
                att = att.iloc[iatt - iatt1]
                att.columns = hdr_data.iloc[0]
                att_trans = att.transpose().reset_index()
                att_trans.columns = att_trans.iloc[0]
                att_trans = att_trans[1:]
                att_trans.columns.values[0] = 'Variable'
            else:
                print(f'No attributes found for Table {ntbl}/{ntables}')
                att_trans = None

        # Read data:
        if ndata > 0:
            tf_tbl = pd.read_csv(flin, header=None, sep=',', skiprows=istart, nrows=ndata, usecols=hdr_select,
                                 keep_default_na=False)
            tf_tbl_codes = tf_tbl[0].str[:3]
            icmt = tf_tbl_codes == 'CMT'
            ncmt = icmt.sum()
            if ncmt > 0:
                print(f'Removing {ncmt} comments from table {tbl_str}')
                tf_tbl = tf_tbl[~icmt]
            tf_tbl.drop(columns=0, inplace=True)
            tf_tbl.columns = hdr_data.iloc[0]
            # reset timezones for all dates:
            for col in tf_tbl.columns:
                rtn = get_attribute_check(att_trans, col, 'Units')
                if rtn is None or pd.isnull(rtn['result']):
                    continue
                if tf_tbl[col].dtype == 'datetime64[ns]':
                    tzone = re.match(r'Time\s*\(\s*(.*)\s*\)', rtn['result'])
                    if tzone:
                        tzone = tzone.group(1)
                        if tzone in all_timezones:
                            print(f'Found valid time and timezone ({tzone}) for column {col}')
                            tf_tbl[col] = tf_tbl[col].dt.tz_localize(get_localzone()).dt.tz_convert(tzone)
                        else:
                            print(f'Invalid timezone ({tzone}) for column {col}, leave as is')
                    elif rtn['result'] in all_timezones:
                        tzone = rtn['result']
                        print(f'Inferred time, found timezone ({tzone}) for column {col}')
                        tf_tbl[col] = tf_tbl[col].dt.tz_localize(get_localzone()).dt.tz_convert(tzone)
                    else:
                        print(f'WARNING: No timezone found for variable {col}, leave as is')
                else:
                    continue
        else:
            tf_tbl = None

        # Add data to result:
        if tf_tbl is None:
            result[tbl_str] = None
        elif tbl_str.lower() == 'global':
            tb_gbl = tf_tbl.transpose().reset_index()
            result[tbl_str] = tb_gbl.rename(columns={0: 'Variable'})
        else:
            result[tbl_str] = tf_tbl

        if att_trans is None:
            result['tc_' + tbl_str] = None
        else:
            result['tc_' + tbl_str] = att_trans

        # Summary message:
        print(f'Tables {tbl_str} and tc_{tbl_str} for table {ntbl}/{ntables}: {tf_description[iheader]}')
        if ndata > 0:
            print(f'  Rows: header at {iheader}, data from {istart} to {iend}')
        elif att_trans is not None:
            print(f'  Rows: header at {iheader}, metadata only from {iatt1} to {iatt2}')
        else:
            print(f'  Rows: header only at {iheader}, no data, no metadata')

    print(f'Finished reading file: {flin}')
    return {'result': result, 'error': None}
ebs_read_tbe()